### PR TITLE
ENH Looping through arrays in templates

### PR DESCRIPTION
**Pull request by GuySartorelli**
## Description

Allows arrays to be passed into templates and iterated over.
Associative arrays are wrapped in `ArrayData` so their values can be accessed by key.
Indexed arrays are wrapped in ~~`ArrayIterator`~~ `ArrrayList` when fetched from `ViewableData::obj()` to respect the method signature.

~~Not wrapping in `ArrayList` because that currently throws exceptions if indexed arrays are passed in, and if we changed that it would silently fail if you try to call `Filter()` or `Sort()` or `Find()` etc.~~

## Issues

- #11237

### CODE REVIEW

File: src/View/ViewableData.php [hunk 1]
```diff
@@ -668,17 +679,6 @@ public function getViewerTemplates($suffix = '')
...
-    /**
-     * When rendering some objects it is necessary to iterate over the object being rendered, to do this, you need
-     * access to itself.
-     *
-     * @return ViewableData
-     */
-    public function Me()
-    {
-        return $this;
-    }
```
Comment (GuySartorelli) [inline src/View/ViewableData.php]: There's no reason to ever call this from PHP code since you already have the object itself at that point.

This has been moved into the templating layer. Without this change, using `<% loop $Me %>` with an `ArrayIterator` or `array` in scope will result in trying to call a `Me()` method on it - which obviously throws errors.

File: src/View/ViewableData.php [hunk 2]
```diff
@@ -537,7 +537,7 @@ protected function objCacheClear()
      * @param array $arguments
      * @param bool $cache Cache this object
      * @param string $cacheName a custom cache name
-     * @return Object|DBField
+     * @return object|DBField
```
Comment (GuySartorelli) [inline src/View/ViewableData.php:541]: It's probably the case that this used to mean the old SS3 `Object` which later became `SS_Object` which is effectively `ViewableData` now from a typing perspective.

That said, the way this is implemented, it really can return any object, so I'm going to treat this as though it was just a capital letter away from being correct initially.

File: src/View/SSViewer_DataPresenter.php [hunk 3]
```diff
@@ -172,6 +172,10 @@ public function getInjectedValue($property, array $params, $cast = true)
         // Get source for this value
         $result = $this->getValueSource($property);
         if (!array_key_exists('source', $result)) {
+            // $Me is a special property. If nothing is providing an override, return the current item.
+            if ($property === 'Me') {
+                return ['obj' => $this->getItem()];
+            }
```
Comment (GuySartorelli) [inline src/View/SSViewer_DataPresenter.php:178]: I originally was going to have this return early - but I think it's worth allowing people to implement their own `Me` method or template provider in case someone's doing something really funky and custom.

Ultimately the reason we're moving this here at all is to handle cases where the current item isn't an instance of `ViewableData`, so anything people are doing on their own classes won't be affected either way.

File: src/View/SSViewer_DataPresenter.php [hunk 4]
```diff
@@ -172,6 +172,11 @@ public function getInjectedValue($property, array $params, $cast = true)
         // Get source for this value
         $result = $this->getValueSource($property);
         if (!array_key_exists('source', $result)) {
+            $obj = $this->getItem();
+            // $Me is a special property. If nothing is providing an override, return the current item.
+            if ($property === 'Me' && !isset($obj->$property)) {
+                return ['obj' => $this->getItem()];
+            }
```
Comment (GuySartorelli) [inline src/View/SSViewer_DataPresenter.php]: I originally was going to have this return early - but I think it's worth allowing people to implement their own `Me` method or template provider in case someone's doing something really funky and custom.

Ultimately the reason we're moving this here at all is to handle cases where the current item isn't an instance of `ViewableData`, so anything people are doing on their own classes won't be affected either way.

File: src/View/SSViewer_DataPresenter.php [hunk 5]
```diff
@@ -313,7 +341,14 @@ public function __call($name, $arguments)
...
             if ($name === 'hasValue') {
-                $result = ($obj instanceof ViewableData) ? $obj->exists() : (bool)$obj;
+                // Check if a value exists, e.g. <% if $obj %>
+                if ($obj instanceof ViewableData) {
+                    $result = $obj->exists();
+                } elseif (is_countable($obj)) {
+                    $result = count($obj) > 0;
+                } else {
+                    $result = (bool) $obj;
+                }
```
Comment (GuySartorelli) [inline src/View/SSViewer_DataPresenter.php]: This change allows us to use `<% if $MyArray %>` the same way we'd use `<% if $MyArrayList %>`. It's quite common to rely on this shortcode when checking if a list has values, so it will be intuitive to do it with non-ViewableData iterables as well.

I've made the same change in `ViewableData::hasValue()` for the same reason.

File: src/View/SSTemplateParser.php [hunk 6]
```diff
@@ -1886,6 +1888,8 @@ function PresenceCheck_Argument(&$res, $sub)
             $res['php'] .= '((bool)'.$sub['php'].')';
         } else {
             $php = ($sub['ArgumentMode'] == 'default' ? $sub['lookup_php'] : $sub['php']);
+            // TODO: kinda hacky - maybe we need a way to pass state down the parse chain so
+            // Lookup_LastLookupStep and Argument_BareWord can produce hasValue instead of XML_val
```
Comment (GuySartorelli) [inline src/View/SSTemplateParser.php]: This and other comments in here were already in the `.peg` file and were automatically added here when running the peg file to generate this one. This file has not been manually altered at all.

File: src/View/SSTemplateParser.peg [hunk 7]
```diff
@@ -287,6 +287,8 @@ class SSTemplateParser extends Parser implements TemplateParser
         if (isset($sub['Call']['CallArguments']) && isset($sub['Call']['CallArguments']['php'])) {
             $arguments = $sub['Call']['CallArguments']['php'];
             $res['php'] .= "->$method('$property', [$arguments], true)";
+        } elseif ($property === 'Count') {
+            $res['php'] .= "->$property()";
```
Comment (GuySartorelli) [inline src/View/SSTemplateParser.peg]: Allows us to handle `Count` in a special way in the `__call()` method of the data presenter layer.
This changes the template code from

```php
$val .= $scope->locally()->obj('Iterable', null, true)->XML_val('Count', null, true);
```

to 

```php
$val .= $scope->locally()->obj('Iterable', null, true)->Count();
```

File: src/View/SSViewer_DataPresenter.php [hunk 8]
```diff
@@ -303,6 +308,29 @@ public function getObj($name, $arguments = [], $cache = false, $cacheName = null
...
+                } elseif (isset($item->$name)) {
+                    $result = $item->$name;
+                }
+            } elseif (is_countable($item)) {
+                // Count countables
+                $result = count($item);
+            }
+            $this->resetLocalScope();
+            return $result;
+        }
```
Comment (GuySartorelli) [inline src/View/SSViewer_DataPresenter.php:332]: This allows the use of `$Count` in templates to work the same for arrays and other iterators as it would for any `SS_List` implementation.

It still respects type casting in `ViewableData` when the current item in scope (the item this is being called on) is an instance of `ViewableData`.

File: src/View/SSViewer_DataPresenter.php [hunk 9]
```diff
@@ -303,6 +308,29 @@ public function getObj($name, $arguments = [], $cache = false, $cacheName = null
...
```
Comment (GuySartorelli) [inline src/View/SSViewer_DataPresenter.php]: This allows the use of `$Count` in templates to work the same for arrays and other iterators as it would for any `SS_List` implementation.

It still respects type casting in `ViewableData` when the current item in scope (the item this is being called on) is an instance of `ViewableData`.

File: src/View/ArrayData.php [hunk 10]
```diff
@@ -77,7 +77,7 @@ public function toMap()
     public function getField($field)
     {
         $value = $this->array[$field];
-        if (is_object($value) && !$value instanceof ViewableData) {
+        if (is_object($value) && !($value instanceof ViewableData) && !is_iterable($value)) {
             return new ArrayData($value);
```
Comment (GuySartorelli) [inline src/View/ArrayData.php:81]: This is needed because otherwise `ArrayIterator` and other iterators get shoved inside `ArrayData` which is obviously not correct and results in not being able to get the count.

Arguably this is a bug fix

File: src/View/ViewableData.php [hunk 11]
```diff
@@ -558,6 +558,17 @@ public function obj($fieldName, $arguments = [], $cache = false, $cacheName = nu
...
+        // Wrap arrays
+        if (is_array($value)) {
+            if (array_is_list($value)) {
+                // Wrap in ArrayIterator to respect method signature
+                $value = new ArrayIterator($value);
+            } else {
+                // Wrap in ArrayData so values can be accessed by key in templates
+                $value = ArrayData::create($value);
+            }
+        }
```
Comment (GuySartorelli) [inline src/View/ViewableData.php]: I've taken a look at the usage of `->obj()` in sink and while a lot of them assume various things, those assumptions either are unaffected or would have already been wrong under certain circumstances anyway.

I'm confident this isn't going to cause any breakage of functionality for anyone, so if I do mention it in the changelog it'll be in passing.

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Can you help me understand what is meant by `Not wrapping in ArrayList because that currently throws exceptions if indexed arrays are passed in, and if we changed that it would silently fail if you try to call Filter() or Sort() or Find() etc.

Just doing a quick test on CMS 5 post merging the [scalars PR](https://github.com/silverstripe/silverstripe-framework/pull/11238)

I can pass this in from a controller

```php
    public function ArrayListTest()
    {
        // note the decending order
        return new ArrayList(['item3', 'item2', 'item1']);
    }
```

And render it in a template

```html
<!-- works - item3, item2, item1 -->
<% loop $ArrayListTest %>
    $Me
<% end_loop %>

<!-- works -->
$ArrayListTest.Count

<!-- though this will never work instead it will "silently fail" - item3, item2, item1
     Sort requires a column -->
<% loop $ArrayListTest.Sort %>
    $Me
<% end_loop %>
```

Seem like the major reason for ArrayIterator is because methods like `.Sort` can't work, because we don't have ArrayData() or another object as the members in the ArrayList, and you'd like a hard error thrown rather than nothing happening to protected against developer error?

Though that seems like a strange reason to not use ArrayList? Because you can already construct an ArrayList full of scalars and use it in a template.  It's just called Sort() on it with no parameters does nothing ... though that's what is supposed to happen?

I'm not sure what's so wrong with ArrayList's at we'd want to complicate things by using ArrayIterator() instead which increases the number of scenarios that need to be coded for / things that could go wrong.

I feel like I'm missing something important here?

**Comment by GuySartorelli** [comment]
> Though that seems like a strange reason to not use ArrayList? Because you can already construct an ArrayList full of scalars and use it in a template. It's just called Sort() on it with no parameters does nothing ... though that's what is supposed to happen?

First of all I didn't say anything about no parameters - whether you pass parameters or not it'll have undefined behaviour and most likely fail silently. The assumption is that you can sort and filter arraylists. Remember this affects PHP-land logic as well, not just template logic.

If we want to wrap indexed arrays in `ArrayList`, we need to change the way `ArrayList`'s iterator works so that it doesn't throw an exception.

We then need to decide what happens if someone calls `sort()`, `filter()`, `find()`, or any other methods that require keys to get values in order to operate on them?
    - Do we throw exceptions for these? That's bad DX. `ArrayList` implements `Sortable` and `Filterable`, so it should be sortable and filterable.
    - Do we allow sorting the underlying array in various ways (alphabetically, numerically, etc)? We then need to make sure the API for that is clear, and we're still left with answering what happens if someone tries to sort by a specific key,
    - Do we allow sorting by _value_? If so we have the same problems as with sorting.
    - Do we just let it fail silently? That's the worst possible solution IMO.

We also have to verify what will happen if you try to feed that into a gridfield, or anywhere else that currently expects `ArrayList` that works the way `ArrayList` works right now.

Basically trying to use `ArrayList` here opens a whole can of worms that I'm not willing to open as part of this issue. That's scope creep if ever I saw it. If you want to use `ArrayList` here you're welcome to open a follow-up card to do that after this PR is merged as a separate piece of work. I'm not doing it as part of this card.

**Comment by emteknetnz** [comment]
> We then need to decide what happens if someone calls sort(), filter(), find(), or any other methods that require keys to get values in order to operate on them? ... We also have to verify what will happen if you try to feed that into a gridfield, or anywhere else that currently expects ArrayList that works the way ArrayList works right now.

It does seem a little odd to be concerned about all these things - as mentioned before in CMS 5 you can currently go `new ArrayList(['item3', 'item2', 'item1']);` - we don't seem to be too concerned about the fact you can't actually use this ArrayList in a GridField right now?

Based on what you're suggesting here seems like we may even want to actually restrict the constructor of ArrayList to disallow a index array scalars entirely since you can't use most of the ArrayList API on them? I am actually open to that idea as it would restrict the number of possible codepaths and scenarios that need to be considered ... though obviously it negates the whole idea of just return and array of scalars in a controller for template consumption.

Maybe this card is just a bad idea? I'm starting to now wonder what ArrayList methods with a list of values can and can't be used in a template. Count can, though sort and friends can't, though are there other that can?  Should be making it so that arrays can have the exact same methods called on to match ArrayList? Feels like by allowing mixed types we're just making things worse... typically we want to go in the direction of being more restrictive to make things more robust e.g. strong typing, this feels like it's going in the wrong direction

**Comment by GuySartorelli** [comment]
> It does seem a little odd to be concerned about all these things - as mentioned before in CMS 5 you can currently go `new ArrayList(['item3', 'item2', 'item1']);` - we don't seem to be too concerned about the fact you can't actually use this ArrayList in a GridField right now?

It's not so much about gridfield (that was just one example) as it is about the API for `ArrayList` not being respected or usable when you throw data into it which doesn't have keys to sort/filter/etc by.
Right now that's not really a problem because as soon as you try to iterate over the list it'll throw an exception at you anyway.

> Based on what you're suggesting here seems like we may even want to actually restrict the constructor of ArrayList to disallow a index array scalars entirely since you can't use most of the ArrayList API on them? I am actually open to that idea as it would restrict the number of possible codepaths and scenarios that need to be considered ... 

That would be great, as a separate issue. This issue isn't about deciding how to fix the problems with arraylist.

> though obviously it negates the whole idea of just return and array of scalars in a controller for template consumption.

No it doesn't, as this PR avoids using `ArrayList` and allows you to return an array of scalars in a controller for template consumption.

> Maybe this card is just a bad idea?

I think this card is a great idea, and I think we can worry about `ArrayList` separately.

> I'm starting to now wonder what ArrayList methods with a list of values can and can't be used in a template. Count can, though sort and friends can't, though are there other that can? Should be making it so that arrays can have the exact same methods called on to match ArrayList? Feels like by allowing mixed types we're just making things worse... typically we want to go in the direction of being more restrictive to make things more robust e.g. strong typing, this feels like it's going in the wrong direction

With this PR, `$Count` will work in a template as an explicit and special variable, because it is common to want to know how many items are in your iterable variable.
Sorting and filtering of regular arrays can be done in a controller - if we want to add a way to do that from within the template that would be a neat enhancement - but we have to be able to use the arrays in a template in the first place before we can look at that.

**Comment by GuySartorelli** [comment]
Do we need to have a meeting to discuss this one or can we get it merged?

**Comment by emteknetnz** [comment]
My main concern here is that by not using ArrayList + ArrayData we're increasing the complexity of the number of scenarios that need to be managed.

As an alternate solution, would it work if arrays were converted right before going into the template to a combination of ArrayList and ArrayData with a single magical "_index" key used in the ArrayData if an index array was passed into ArrayList?

e.g. 

```php
['item1', 'item2'];
```

Automatically gets converted to

```php
new ArrayList([
  new ArrayData['_index' => 'item1'],
  new ArrayData['_index' => 'item2'],
]);
```

And then have some logic where if the only key in the ArrayData is `_index` then use that for the value of `$Me`?

Seems like this would negate the need for all the extra logic to handle iterables / countables etc?

**Comment by GuySartorelli** [comment]
It's not only templates though, it's also PHP. So if you do that, you're also saying any time you pass a non-associative array to `ArrayList` it'll get this weird key and be wrapped in `ArrayData`. But if you iterate over that `ArrayList` in PHP it won't know magically that you're trying to get the value from the `_index` key  so you have to account for that in your project code, which is a little gross.

If you're really insistent on this being handled using `ArrayList`, as I've mentioned, that opens a whole can of worms that will have to be thoroughly discussed. I don't want to do that, I want to just get something working in an intuitive way which IMO this PR does, and we can worry about `ArrayList` separately.

Looking at this PR again, there's not that much complexity being added to account for non-`ArrayList` iterables.
1. There's a small bugfix in `ArrayData` that IMO should happen anyway.
1. There's moving `$Me` from `ViewableData` into `SSViewer_DataPresenter` which IMO should happen anyway.
1. There's a small change in the parser to allow `$Count` logic to be handled differently.
    - This is just a simple conditional statement.
1. There's handling `$Count` in `SSViewer_DataPresenter`, which is probably the biggest piece of added complexity. It's 22 lines including comments, and all it does is
    1. If `ViewableData`, Call `XML_val('Count')`
    1. If there's a method or property called `Count` on the item, use that
        - This isn't strictly necessary, we could remove this if we're that concerned about the complexity but IMO it's not adding much maintenance burden.
    1. if the item is countable , use `count($item)`
    1. Everything else returns `null`
1. There's some logic in both `ViewableData` and `SSViewer_DataPresenter` to allow an `<% if $MyIterable %>` to check the count of the iterable.
    - This is just a simple conditional statement.
    - We could move this logic into a single place if we're concerned about the duplication - but there's quite a few spots like this between `SSViewer` (and its scope) and `ViewableData` so we would probably want to look at that deduping separately.
1. There's the logic in both `ViewableData` and `SSViewer_Scope` to wrap the iterable, which has to happen regardless of what we're wrapping it in
1. There's a tiny condition check for `$this->item instanceof Iterable`

Choosing to wrap this in `ArrayList` will remove 3, 4, and 5 above. The complexity of these is discussed above.
It will also probably add:

1. Changing `ArrayList` so it doesn't throw exceptions when iterating over a non-associative array
3. Possibly some new logic in sort/filter/find/etc to handle what happens with non-associative arrays
4. Possibly some logic either in `ArrayData` or somewhere in the templating system to check if there's a `_index` key in the item we're trying to:
    1. Call `$Me` to get the current item
    1. access a property or call a method (not just `$Me`, we need to be able to access any method or property on the item) - (currently handled by `obj()` on `ArrayData` so we're changing the way that works in PHP land as well), or
    4. check if it has a value (currently handled by `hasValue()` on `ArrayData` so we're changing the way that works in PHP land as well), or
    6. iterate over it (maybe this would go in `getItem()` or `next()` on `SSViewer_Scope`? Not sure), or
    1. Other scenarios? Are there edge cases we're missing that will result in new bugs we have to deal with down the line?


I think we'd quickly find trying to deal with `ArrayList` here becomes more complicated very quickly, and has much wider reaching changes than the ones I've made in this PR and therefore a higher chance of causing regressions or future unexpected behaviours.

**Comment by GuySartorelli** [comment]
So again.
> Do we need to have a meeting to discuss this one or can we get it merged?

**Comment by emteknetnz** [comment]
Probably have a chat to discuss

I've quickly together a PR to demo the idea of just using an ArrayList https://github.com/silverstripe/silverstripe-framework/pull/11256, seems like it does the same thing with much less code changes required? (I've probably missed something though)

File: src/View/ViewableData.php (hunk 2 above)
Comment (GuySartorelli) [reply src/View/ViewableData.php:541]: Update now that this PR uses `ArrayList` instead of `ArrayIterator`:
There are still other scenarios where this method could return arbitrary objects, because of the way it's been implemented. It might make more sense to tighten this up to require `ViewableData` to be returned but for now that's just not how it works - and doing that should be handled in a separate PR.

File: src/View/ViewableData.php (hunk 1 above)
Comment (GuySartorelli) [reply src/View/ViewableData.php]: Update now that this PR uses `ArrayList` instead of `ArrayIterator`:
I still think it's worth doing this - if you want me to split it out into a separate issue and PR I will but I'm leaving it here for now.

**Comment by GuySartorelli** [comment]
We've discussed offline - I had a faulty assumption about `ArrayList` - we'll try with `ArrayList` instead of `ArrayIterator`

File: src/ORM/ArrayList.php [hunk 12]
```diff
@@ -127,7 +127,7 @@ public function exists()
     public function getIterator(): Traversable
     {
         foreach ($this->items as $i => $item) {
-            if (is_array($item)) {
+            if (is_array($item) && !array_is_list($item)) {
```
Comment (GuySartorelli) [inline src/ORM/ArrayList.php:130]: Ultimately the way `ArrayList` works needs to be re-evaluated at some stage anyway - only converting in the iterator but not when fetching items directly via `find()` and other methods leads to unexpected results. So I think this is fine for now even though it allows a setup that used to cause an explicit error.

File: src/View/SSViewer_Scope.php [hunk 13]
```diff
@@ -308,6 +320,8 @@ public function next()
             // Item may be an array or a regular IteratorAggregate
             if (is_array($this->item)) {
                 $this->itemIterator = new ArrayIterator($this->item);
+            } elseif ($this->item instanceof Iterator) {
+                $this->itemIterator = $this->item;
             } else {
```
Comment (GuySartorelli) [inline src/View/SSViewer_Scope.php:320]: I think it's worth adding this because:
1. there's already logic here to handle native arrays, even though previously those couldn't be used at all and now they're wrapped in `ArrayList` before we get here... but this allows for them to be handled correctly if they somehow sneak in
2. As with the array check, this allows iterators to be used correctly if they somehow sneak in. They shouldn't get here, but if they do it's better to just let them be iterated over rather than throw exceptions about it, IMO.

File: src/View/ViewableData.php [hunk 14]
```diff
@@ -601,7 +612,10 @@ public function cachedCall($fieldName, $arguments = [], $identifier = null)
     public function hasValue($field, $arguments = [], $cache = true)
     {
         $result = $this->obj($field, $arguments, $cache);
-        return $result->exists();
+        if ($result instanceof ViewableData) {
+            return $result->exists();
+        }
+        return (bool) $result;
```
Comment (GuySartorelli) [inline src/View/ViewableData.php:613]: This is still needed because as mentioned above the `obj()` method can return arbitrary objects with its current implementation. It's not related to the issue at hand but worth keeping IMO.

File: tests/php/View/SSViewerTest.php [hunk 15]
```diff
@@ -1016,6 +1017,48 @@ public function testIfBlocks()
...
+                'inScope' => false,
+            ],
+            'array' => [
+                'iterable' => [1, 2, 3],
+                'inScope' => false,
+            ],
+            'iterator' => [
+                'iterable' => new ArrayIterator([1, 2, 3]),
+                'inScope' => false,
+            ],
```
Comment (GuySartorelli) [inline tests/php/View/SSViewerTest.php]: I'm surprised this works - but I think it's worth keeping it working, since it doesn't require much additional complexity to maintain.

File: tests/php/View/SSViewerTest.php [hunk 16]
```diff
@@ -1331,6 +1374,88 @@ public function testCastingHelpers()
...
+            'nested array and iterator' => [
+                'iterable' => [
+                    [
+                        'value 1',
+                        'value 2',
+                    ],
+                    new ArrayIterator([
+                        'value 3',
+                        'value 4',
+                    ]),
```
Comment (GuySartorelli) [inline tests/php/View/SSViewerTest.php]: I'm surprised this works - but I think it's worth keeping it working, since it doesn't require much additional complexity to maintain.

**Review by GuySartorelli — COMMENTED** [review]
Some of the added tests are no longer explicitly required for the new functionality - but they weren't being tested before so we're adding additional (if only tangentially related to the current issue) test coverage.

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Code changes look good, though this one should be run though a kitchen-sink CI run with the PR included since it's a pretty low level change

One thing I did notice is that while `<% loop $Me %>` worked as expected `<% loop %>` did not, possibly the PR just needs a rebase or merge-up though

Test code:

**PageController**
```php

    public function OneList()
    {
        return ['abc', 'def'];
    }

    public function OneAssoc()
    {
        return ['Foo' => 'bar'];
    }

    public function TwoList()
    {
        return [
            ['abc', 'def'],
            ['ghi', 'xyz'],
        ];
    }

    public function TwoAssoc()
    {
        return [
            [
                'Food' => 'bars',
                'Bazz' => 'buzz',
            ],
        ];
    }

    public function ThreeList()
    {
        return [
            [
                ['123', '456'],
            ],
            [
                ['789', '000'],
            ]
        ];
    }

    public function ThreeAssoc()
    {
        return [
            [
                [
                    'Foody' => 'barsy',
                    'Bazzy' => 'buzzy',
                ],
            ],
        ];
    }
```

**Page.ss**
  ```html
    <% loop $OneList %>
        $Me<br>
    <% end_loop %>

    $OneAssoc.Foo<br>

    <% loop $TwoList %>
        <% loop $Me %>
            $Me<br>
        <% end_loop %>
    <% end_loop %>

    <% loop $TwoAssoc %>
        $Me.Food<br>
        $Me.Bazz<br>
    <% end_loop %>

    <% loop $ThreeList %>
        <% loop $Me %>
            <% loop $Me %>
                $Me<br>
            <% end_loop %>
        <% end_loop %>
    <% end_loop %>

    <% loop $ThreeAssoc %>
        <% loop $Me %>
            $Me.Foody<br>
            $Me.Bazzy<br>
        <% end_loop %>
    <% end_loop %>
```

**Comment by GuySartorelli** [comment]
> One thing I did notice is that while `<% loop $Me %>` worked as expected `<% loop %>` did not, possibly the PR just needs a rebase or merge-up though

This PR started before that work got done, so yeah that'd need to be merged up to `6` and then this would need to be rebased on top.
I'm not gonna bother doing that unless you explicitly ask for it because that doesn't really feel related to this work tbh.

**Comment by GuySartorelli** [comment]
Had to rebase in the end to get kitchen sink CI to run happily. That CI run is linked in the issue.

**Review by GuySartorelli — COMMENTED** [review]
Regrettably, the `Me` deprecation in `ViewableData` caused other problems in sink, because against all reason it's being used in PHP contexts.

I've removed that from this PR and have [opened a new issue](https://github.com/silverstripe/silverstripe-framework/issues/11261) for it.

That has also stopped the `ArrayIterator` tests in `SSViewerTest` from working (because there's no `Me()` method on that class) so I've removed those tests as well.

**Comment by GuySartorelli** [comment]
Wrapping arrays in `ArrayData` means it's no longer being cast with the old casting logic, which is causing problems in sink.
This points to a wider problem with how that casting is being done or relied on.

I've [opened a separate card](https://github.com/silverstripe/silverstripe-framework/issues/11262) to look into all that but for now I've removed the logic wrapping arrays in `ArrayData` from this PR. We're now explicitly only dealing with list arrays.

**Review by emteknetnz — APPROVED** [review]
Tested locally, works well
